### ISSUE TITLE
ONNX conversion to TensorRT with trtexec successfull but polygraphy TensorRT inference error, TensorRT 8.6 when running on Windows 10 on GTX 1660 Ti

### ISSUE THREAD

**Issue by ninono12345** [NVIDIA/TensorRT#3578]
## Description

  
Hello everyone willing to help out. I am a student, my professor gave me some extra work, that is to take
[pytracking framework](https://github.com/visionml/pytracking) using the tomp tracker, to convert the model to TensorRT so that it could be inferred faster.

Keep in mind that I am using Windows 10, and I am not using any environment, just straight up installed python 3.10 on my pc and pip installed all the packages, CUDA 12.1 update 1

Now if you look at the code it is a mess, the models used for tracking do not have a simple forward function, everything is a mess. All model data is inside the tomp101.pth file, but for everything to work I separated it into 2 models, so I rewrote small parts of the code, and now there are 2 models that this tracker uses - one for initialization and the second one for keeping track of the selected object. Everything works the same, but now there are simple forward functions so we can convert the models to onnx and tensorrt

I am using the tomp101 model, so to run the tracker with my modified forward functions write: python run_webcam.py tomp tomp101

All model data is inside the tomp101.pth.tar file, but for everything to work I separated 

I convert both models to onnx:

inside: D:\pyth\pytracking-master\pytracking\tracker\tomp\tomp.py

line 23:
`do_train = True`
`do_conversion_onnx_tompnet = False`
`do_conversion_onnx_head = False`

--------

line 420:
`tm = self.net.net
torch.onnx.export(tm,
                        (im_patches.to("cuda")),
                        "tomp101_tompnet_new.onnx",
                        verbose=False,
                        export_params=True,
                        do_constant_folding=True,
                        opset_version=16,
                        input_names=['im'],
                        output_names=['backbone_feat']
                        ,dynamic_axes={'im':{0:'batch_size'},
                                    'backbone_feat':{0:'batch_size'}})`

this has one simple input shape (1,3,288,288) and output shape (1,1024,18,18)

and

`hn = self.net.net.head
torch.onnx.export(hn,

        model_inputs,
        "tomp101_head_latest.onnx",
        verbose=False,
        export_params=True,
        do_constant_folding=True,
        opset_version=16,
        input_names=input_names,
        output_names=output_names,
        dynamic_axes={'sample_x':{0:'batch_size'},
                    'train_samples':{0:'batch_size'},
                    'target_labels':{0:'batch_size'},
                    'train_ltrb':{0:'batch_size'},
                    'target_scores':{0:'batch_size'},
                    'bbox_preds':{0:'batch_size'}})`

input shapes: 1. (1,1024,18,18) 2. (1,1024,18,18) 3. (1,1,18,18) 4. (1,4) - but for some reason the 4th input converts into (1,4,18,18)
output shapes: 1. (1,1,18,18) 2. (1,1,4,18,18)
yes, the second module is a head of the same model.

The FIRST MODEL, used for initialization, is quite simple, uses only the backbone of ResNet101 and a few more things, so when I convert it to onnx, then run test the converted model with [polygraphy](https://github.com/onnx/onnx-tensorrt/blob/main/docs/faq.md) and how would it convert to tensorrt: polygraphy run tomp101_tompnet_new.onnx --trt everything goes well

Here is the definition of the SECOND MODELS forward function:

inside: D:\pyth\pytracking-master\ltr\models\transformer\heads.py

line 61:
START OF CODE (as you can see gitgub code formatting is bad)
`
def forward(self, sample_x, train_samples=None, target_labels=None, train_ltrb=None):

    """classify target by applying the dimp filter."""
    with torch.no_grad():
        print("we are use_head_forward")
        print(type(sample_x))
        test_feat = self.extract_head_feat(sample_x)
        train_feat = self.extract_head_feat(train_samples)
        cls_weights, bbreg_weights, cls_test_feat_enc, bbreg_test_feat_enc = \
            self.get_filter_and_features_in_parallel(train_feat, test_feat, num_gth_frames=1,
                                                              train_label=target_labels, train_ltrb_target=train_ltrb)

        # fuse encoder and decoder features to one feature map
        target_scores = self.classifier(cls_test_feat_enc, cls_weights)

        # compute the final prediction using the output module
        bbox_preds = self.bb_regressor(bbreg_test_feat_enc, bbreg_weights)

    return target_scores, bbox_preds`
END OF CODE

The second model, when testing with Polygraphy: polygraphy run tomp101_head_latest.onnx --trt, gives out this error:

`[E] C:\_src\plugin\instanceNormalizationPlugin\instanceNormalizationPlugin.cu (335) - Cudnn Error in nvinfer1::plugin::InstanceNormalizationPlugin::enqueue: 14 (CUDNN_STATUS_VERSION_MISMATCH)`

Now there are 2 things that as much research as I did, I could not find the answer to:

1. When running on a docker NGC pytorch container, this second model seems to work just fine. One could argue that maybe I haven't installed cuDNN correctly on my windows system, but I have read through cuDNN installation guide 10 times, so I don't think that bad installation is a problem, but this brings me to the second problem:
2. The simpler first model went through polygraphy test just fine, so this means that TensorRT is installed correctly, everything works just fine, only the second model is problematic, or mabye there are some incompatabilities, when inferencing it on Windows 10.

Running the app on docker is not an option, because of dockers limitations with webcam capture with opencv

So now I either have to somehow recreate everything that is in the docker environment on my Windows 10, because somehow there everything works...
Or maybe I have to modify the second model, because if the first one works, so can the second, right?

One thing that's caught my eye, if we upload the second ONNX model into NETRON, it shows the output shapes as such:

target_scores: tensor: float32[batch_size,1,Reshapetarget_scores_dim_2,18]
bbox_preds: tensor: float32[batch_size,Unsqueezebbox_preds_dim_1,4,Unsqueezebbox_preds_dim_3,Unsqueezebbox_preds_dim_4]

Does this have any effect on the output of the model, or if it would work overall?

I have run on the second onnx model [polygraphy fold constants](https://github.com/NVIDIA/TensorRT/tree/main/tools/Polygraphy/examples/cli/surgeon/02_folding_constants), but this doesn't seem to change anything

Also when converting onnx to tensorrt via trtexec I get no errors whatsoever, but as I understand the polygraphy run model.onnx --trt also tries to do inference on the model and that's where it fails

on the left is the same model run on Windows system, on the right through nvidia pytorch 23.12 NGC docker:

![tompnet_head](https://github.com/NVIDIA/TensorRT/assets/116962708/b5071c38-d3a4-4373-b8f6-0d8c5cda4f63)

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[NETRON graph](https://drive.google.com/file/d/1jLoDqh40WtqVMkAJxARqWpiQ2CfLX2Ha/view?usp=sharing)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Anybody that has some experience with this please help me, because I've been stuck on this for a month, I was making very very slow progress, but now it's like I've hit a wall and I have no idea what to do.

I am attaching my source code as well as instructions on how i converted my models to onnx bellow

IF ANY ADDITIONAL INFO IS REQUIRED PLEASE DO TELL

THANK YOU FOR YOUR HELP!!!

## Environment

**TensorRT Version**: 8.6

**NVIDIA GPU**: GTX 1660 Ti

**NVIDIA Driver Version**: 545.29.01

**CUDA Version**: 12.1 update 1

**CUDNN Version**: 8.9.7

Operating System: Windows 10

Python Version: 3.10.13

PyTorch Version: 2.1.2+cu121

Runing straight from windows, no environment

## Relevant Files

**Model link**: [my google drive](https://drive.google.com/drive/folders/1Unj7DZi6ACGsUVmSunDPM2WcnvRXBHrv)

## Steps To Reproduce

go to my google drive: [ my drive](https://drive.google.com/drive/folders/1Unj7DZi6ACGsUVmSunDPM2WcnvRXBHrv)

Download 4mb pytracking-master and separately the model and the converted onnx models.
put tomp101.pth.tar inside pytracking-master\pytracking\networks\
OR
download 800mb file with everything. Converted ONNX models will be inside pytracking-master\pytracking

The pytracking master MUST BE at the exact directory D:\pyth\pytracking-master\ or NOTHING WILL WORK!!!!

**Commands or scripts**:

`polygraphy run tomp101_tompnet_new.onnx --trt` - everything works fine
`trtexec --onnx=tomp101_tompnet_new.onnx` - everything also works fine

`polygraphy run tomp101_head_latest.onnx --trt` - inference error
`trtexec --onnx=tomp101_head_latest.onnx` -- everything works fine...

**Comment by zerollzeng** [comment]
I can help you solve the question about TensorRT but I cannot help you with the model, I quickly went through your description and could you please describe your question about TensorRT? do you find any bug or unexpected behavior on TRT?

**Comment by ninono12345** [comment]
> I can help you solve the question about TensorRT but I cannot help you with the model, I quickly went through your description and could you please describe your question about TensorRT? do you find any bug or unexpected behavior on TRT?

Hello, @zerollzeng, thank you very much for your help, I am very sory for the late reply.

With TensorRT I have minimal experience, I have only been successful in using torch-tensorrt on the docker, but after looking through the internet I couldn't find an explanation on how to install torch-tensorrt on windows.
Only success I've had is with the [torch2trt](https://github.com/NVIDIA-AI-IOT/torch2trt) library. In this case the first model converts and runs, but the second one doesn't
Also I wasn't yet able to successfully run inference on any model when converting model -> onnx -> trtexec, probably missed something in the NVIDIA documentation.

So to answer your question I couldn't 

The help to change the code in the model which I specified is not needed, because using NETRON I found out what was causing the problem.

According to this error:

`[E] C:\_src\plugin\instanceNormalizationPlugin\instanceNormalizationPlugin.cu (335) - Cudnn Error in nvinfer1::plugin::InstanceNormalizationPlugin::enqueue: 14 (CUDNN_STATUS_VERSION_MISMATCH)`

I searched for InstanceNormalization in NETRON and found out that one of the neural networks that our model uses, does InstanceNormalization. I will upload the NETRON graph, you can type in search Instance normalization, to see in more detail.

In this code the problem seems to be centered at self.tower:

`class DenseBoxRegressor(nn.Module):

    def __init__(self, num_channels, project_filter=True):
        super().__init__()
        self.num_channels = num_channels
        self.project_filter = project_filter

        if self.project_filter:
            self.linear = nn.Linear(self.num_channels, self.num_channels)

        layers = []
        layers.extend(conv_layer(num_channels, num_channels))
        layers.extend(conv_layer(num_channels, num_channels))
        layers.extend(conv_layer(num_channels, num_channels))
        layers.extend(conv_layer(num_channels, num_channels))
        self.tower = nn.Sequential(*layers)

        self.bbreg_layer = nn.Conv2d(num_channels, 4, kernel_size=3, dilation=1, padding=1)

    def forward(self, feat, filter):
        nf, ns, c, h, w = feat.shape

        if self.project_filter:
            filter_proj = self.linear(filter.reshape(ns, c)).reshape(ns, c, 1, 1)
        else:
            filter_proj = filter

        attention = filter_layer.apply_filter(feat, filter_proj) # (nf, ns, h, w)
        feats_att = attention.unsqueeze(2)*feat # (nf, ns, c, h, w)

        feats_tower = self.tower(feats_att.reshape(-1, self.num_channels, feat.shape[-2], feat.shape[-1])) # (nf*ns, c, h, w)

        ltrb = torch.exp(self.bbreg_layer(feats_tower)).unsqueeze(0) # (nf*ns, 4, h, w)
        return ltrb`

When taking out this small neural network out of our model, the model now goes through the polygraphy test with no problem:

![image](https://github.com/NVIDIA/TensorRT/assets/116962708/e7d3629a-7292-4514-9ce2-70ac53701ac0)

In theory I could have this model as is, and have the small nn run outside of tensorrt in pytorch, but in the future, there might not be such an option, so I will take any help you can provide.

Thank you again, and I apologize for a late reply.

**Comment by ninono12345** [comment]
Hello, @zerollzeng, so like you mentioned, could you help solve this problem with the error code:
`[E] C:\_src\plugin\instanceNormalizationPlugin\instanceNormalizationPlugin.cu (335) - Cudnn Error in nvinfer1::plugin::InstanceNormalizationPlugin::enqueue: 14 (CUDNN_STATUS_VERSION_MISMATCH)` ?

Thank you

**Comment by zerollzeng** [comment]
> `[E] C:\_src\plugin\instanceNormalizationPlugin\instanceNormalizationPlugin.cu (335) - Cudnn Error in nvinfer1::plugin::InstanceNormalizationPlugin::enqueue: 14 (CUDNN_STATUS_VERSION_MISMATCH)` ?

Looks like you are using a incompatible cuDNN version, I guess it's too old. maybe try install a newer one.

**Comment by zerollzeng** [comment]
> With TensorRT I have minimal experience, I have only been successful in using torch-tensorrt on the docker, but after looking through the internet I couldn't find an explanation on how to install torch-tensorrt on windows.

You can file issue against torch-trt and ask for help.

**Comment by ninono12345** [comment]
@zerollzeng So an update, I have successfully converted the model from onnx to tensorrt and run it, It runs without error, but does have accuracy issues, and I believe it is because int64 values are clamped to int32.
I think this issue can be fixed by applying quantization on the model, before conversion. If anybody has any suggestions, please let me know

`[E] C:\_src\plugin\instanceNormalizationPlugin\instanceNormalizationPlugin.cu (335) - Cudnn Error in nvinfer1::plugin::InstanceNormalizationPlugin::enqueue: 14 (CUDNN_STATUS_VERSION_MISMATCH)`
I am still unable to fix this error. The cuDNN version is 8.9.7, I've also tried installing the 8.9.0, like it says in TensorRT full developers guide, didn't work, then tried installing 8.9.7 again, still getting the same error.

Like I mentioned, for some reason it doesn't like the Instance Normalization that is happening in one of the connected model to the bigger model, maybe there is something else that I missed, or maybe it is an issue with Windows?

Thank you

**Comment by ninono12345** [comment]
fixed when using parser.set_flag(trt.OnnxParserFlag.NATIVE_INSTANCENORM) , to do this we are unable to use the polygraphys converter
